### PR TITLE
Phpcr cleaner command

### PR DESCRIPTION
**Pull request by mamazu**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | -
| Related issues/PRs | -
| License | MIT
| Documentation PR | -

#### What's in this PR?

This PR adds a command to clean dead nodes from the phpCR.

#### Why?

Sulu doesn't delete properties when the data is no longer needed in the documents. This can pose a problem if you want to do a migration on phpcr nodes. You get a lot of unused properties that make it hard to distinguish if migrations are correct and it was just dead properties.

It also reduces the storage size.

#### Example Usage

```bash
bin/console sulu:document:clean
```

#### Known problems

With a lot of nodes and a lot of properties this command can run for a long time and even run into memory issues. For more info see this MR: https://github.com/jackalope/jackalope-doctrine-dbal/pull/421

#### To Do

- [x] Create a documentation PR
- [x] Add breaking changes to UPGRADE.md

#### Open Issues:

1. Shadowed Property

* Change a Block property to a text_line
* Block content will be the content of the new property
* The text_line content is a shadowedProperty value

2. Root properties

* When a root property was removed
* This data will not be removed

### CODE REVIEW

**Comment by mamazu** [comment]
Should I also write some documentation on how to do migrations with the PropertyParser on a lower level? Or is this even worth it, if phpcr gets removed in the next major version?

**Comment by mamazu** [comment]
I'll keep the code but I don't think that this will ever need to be merged.

**Comment by chirimoya** [comment]
@mamazu I would love to have this PR in the 2.6 release. I've asked @wachterjohannes to review and test it last week. Can you reopen the PR?

**Comment by mamazu** [comment]
Sure, let me know if anything needs updating and I'll help you get it merged. I was just cleaning out pull requests. :)

**Comment by chirimoya** [comment]
@mamazu I think @wachterjohannes will get in touch with you, as soon as he has time. Maybe you can already have a look at the  PHP Linting. Thanks!

**Comment by mamazu** [comment]
The tests are broken again. We aren't handling what I called "shadowed properties". I can add the logic back to ignore it again. But we might need a better solution here.

File: src/Sulu/Bundle/DocumentManagerBundle/Command/PhpCRCleanerCommand.php [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+        private StructureMetadataFactoryInterface $structureMetaDataFactory
+    ) {
+        parent::__construct();
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $this->output = $output;
+        $queryManager = $this->session->getWorkspace()->getQueryManager();
+        $rows = $queryManager->createQuery('SELECT * FROM [nt:unstructured]', 'JCR-SQL2')->execute();
```
Comment (mamazu) [inline src/Sulu/Bundle/DocumentManagerBundle/Command/PhpCRCleanerCommand.php:74]: Maybe could filter for pages here as well, as this thing does not support snippets and reduce the number of errors later.

File: src/Sulu/Component/Export/Export.php [hunk 2]
```diff
@@ -111,7 +111,7 @@ protected function getBlockPropertyData(BlockMetadata $property, $propertyValue)
             $blockType = $blockData['type'];
 
             $block = $this->getPropertiesContentData(
-                $property->getComponentByName($blockType)->getChildren(),
+                $property->getComponentByName($blockType)?->getChildren() ?? [],
```
Comment (wachterjohannes) [inline src/Sulu/Component/Export/Export.php:114]: @mamazu why is this needded?

File: src/Sulu/Bundle/DocumentManagerBundle/Command/PhpCRCleanerCommand.php [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+                /** @var array<mixed>|null $subValue */
+                $subValue = $data[$subKey] ?? null;
+                $this->migrate($subKey, $subValue, $value, $depth);
+            }
+
+            return;
+        }
+
+        if ($metaData instanceof BlockMetadata) {
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/DocumentManagerBundle/Command/PhpCRCleanerCommand.php:160]: This section makes sense but only for Block and ComponentMetadata. But not really for PropertyMetadata right?

File: src/Sulu/Bundle/DocumentManagerBundle/Command/PhpCRCleanerCommand.php [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+        } elseif ($metaData instanceof ComponentMetadata) {
+            if (null === $data) {
+                Assert::eq(0, $depth, 'Empty arrays are only allowed at the root');
+                $this->depthDump('Empty page? ' . $metaData->getName(), $depth, OutputInterface::VERBOSITY_VERY_VERBOSE);
+            } else {
+                $this->handleComponentMetadata($data, $metaData, $depth);
+            }
+        } elseif ($metaData instanceof PropertyMetadata) {
+            if (null === $data) {
+                Assert::eq(0, $depth, 'Empty arrays are only allowed at the root');
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/DocumentManagerBundle/Command/PhpCRCleanerCommand.php]: hmm this is also quite strange why is null or empty arrays are only allowed in the root level

File: src/Sulu/Bundle/DocumentManagerBundle/Command/PhpCRCleanerCommand.php [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+                $subValue = $data[$subKey] ?? null;
+                $this->migrate($subKey, $subValue, $value, $depth);
+            }
+
+            return;
+        }
+
+        if ($metaData instanceof BlockMetadata) {
+            if (null === $data) {
+                Assert::eq(0, $depth, 'Empty arrays are only allowed at the root');
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/DocumentManagerBundle/Command/PhpCRCleanerCommand.php:162]: Null data for blocks seem legit to me! they can be ignored in my opinion (nevermind if they are on root level or not

**Review by wachterjohannes — COMMENTED** [review]
Root level properties wont be cleanedup

**Comment by wachterjohannes** [comment]
@mamazu i have created a MR in your REPO ontop of this one to keep track about the changes

**Comment by mamazu** [comment]
Just putting the link here for easy access: https://github.com/mamazu/sulu/pull/2

**Comment by wachterjohannes** [comment]
closed in favor of #7291

**Pull request closed without merging**